### PR TITLE
 Removes OnValidate in cases where it was being used inappropriately

### PR DESCRIPTION
**Pull request by Railboy**
## Overview

There were several cases where OnValidate was being used inappropriately. This PR either moves those checks into Awake / OnEnable or removes them entirely.

OnValidate should be used to validate settings when standard field and RequireComponent attributes aren't up to the task, and where the same validation in an inspector would be unreasonably complex.

Some cases use OnValidate correctly. They have been marked with a comment.

ExecuteInEditMode has been replaced with [ExecuteAlways](https://docs.unity3d.com/ScriptReference/ExecuteAlways.html) for consistency. Behavior is the same.

## Changes
- Fixes: #4202

### CODE REVIEW

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs [hunk 1]
```diff
@@ -82,6 +77,11 @@ private void CheckAsset(object asset, string fieldname)
 
         public void Update()
         {
+            if (!Application.isPlaying)
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: Can update be called when the app isn't playing?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs [hunk 2]
```diff
@@ -82,6 +77,11 @@ private void CheckAsset(object asset, string fieldname)
 
         public void Update()
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: Do you know if there's a reason all the MonoBehaviour methods are `public` in this class?

File: Assets/MixedRealityToolkit.SDK/Features/Audio/Influencers/AudioInfluencerController.cs [hunk 3]
```diff
@@ -151,7 +151,9 @@ public float NativeHighPassCutoffFrequency
         }
 
         private void Awake()
-        {           
+        {
+            audioSource = GetComponent<AudioSource>();
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/Audio/Influencers/AudioInfluencerController.cs]: For many of these, the field will likely have

```
[SerializeField]
[HideInInspector]
```

which could be removed as well if we're moving to this pattern.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs (hunk 2 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: I don't. I'd prefer if they were private.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs (hunk 1 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: When a class uses the ExecuteAlways attribute Update can get called.

Comment (keveleigh) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: This class doesn't appear to have that attribute though

Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: My mistake, thanks.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Cursors/CursorModifier.cs [hunk 4]
```diff
@@ -243,7 +243,7 @@ void IMixedRealityFocusChangedHandler.OnBeforeFocusChange(FocusEventData eventDa
         #region MonoBehaviour Implementation
 
         private void OnValidate()
-        {
+        {   // This is an appropriate use of OnValidate
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Cursors/CursorModifier.cs]: I'm not super convinced there's a difference between this case and any of the others. In both cases, we're calling `GetComponent`. One just serializes the data while the other checks that it exists.

Serializing data at edit time is a valid thing to do. What's inappropriate about the other uses?

Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Cursors/CursorModifier.cs]: Yeah it's pretty subjective, I could go either way on this one. I think what tipped the scales here was:
- you can't use RequireComponent because you need a _generic_ collider
- a custom inspector seems like total overkill for such a simple check
- this feels like an issue we should raise in-editor? This is the vaguest criteria.

**Comment by Railboy** [comment]
Merged latest changes.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/ParabolicTeleportPointer.cs [hunk 5]
```diff
@@ -26,36 +27,33 @@ public class ParabolicTeleportPointer : TeleportPointer
...
-                parabolicLineData.LineTransform = pointerHelper.transform;
-            }
         }
 
         private void EnsureSetup()
         {
             if (parabolicLineData == null)
             {
-                parabolicLineData = GetComponent<ParabolaPhysicalLineDataProvider>();
+                parabolicLineData = gameObject.GetComponent<ParabolaPhysicalLineDataProvider>();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/ParabolicTeleportPointer.cs]: Noticed this one was bigger than the others -- just making sure since we don't have test coverage here, could you check to make sure teleport pointers behave okay after this change?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs [hunk 6]
```diff
@@ -369,5 +369,13 @@ public override void OnInputUp(InputEventData eventData)
             // Poke pointer should not respond when a button is released or hand is un-pinched
             // It should only dispatch events based on collision with touchables.
         }
+
+        protected override void OnEnable()
+        {
+            base.OnEnable();
+
+            Debug.Assert(line != null, "No line renderer found in PokePointer.");
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: Why does pokepointer need a line renderer? Not sure we want to assert here.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs [hunk 7]
```diff
@@ -6,6 +6,7 @@
 
 namespace Microsoft.MixedReality.Toolkit.Input
 {
+    [ExecuteAlways]
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: This didnt' have execute in edit mode, why did you add ExecuteAlways here?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs [hunk 8]
```diff
@@ -82,6 +78,11 @@ private void CheckAsset(object asset, string fieldname)
 
         public void Update()
         {
+            if (!Application.isPlaying)
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: Why do you need this check here? I don't see why sphere pointer needs to have executealwaysattribute, and therefore why it needs this check.

File: Assets/MixedRealityToolkit.Services/InputSystem/NearInteractionTouchable.cs [hunk 9]
```diff
@@ -76,6 +76,11 @@ public void SetLocalUp(Vector3 newLocalUp)
 
         protected new void OnValidate()
         {
+            if (Application.isPlaying)
+            {   // Don't validate during play mode
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/NearInteractionTouchable.cs]: Why don't we want to validate this one during play mode, but it's okay to do this in the components?

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs (hunk 6 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs]: No idea. My priority was just to move code out of OnValidate.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs (hunk 7 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: To ensure OnEnable / CheckInitialization is called in edit mode. This is a substitute for OnValidate.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs (hunk 8 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/SpherePointerVisual.cs]: This class now uses ExecuteAlways so OnEnable is called in edit mode. Update also gets called, but because it executes more than initialization logic we exit early if we aren't in play mode.

File: Assets/MixedRealityToolkit.Services/InputSystem/NearInteractionTouchable.cs (hunk 9 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.Services/InputSystem/NearInteractionTouchable.cs]: In play mode the values being validated here are automatically checked by the methods you use to set them (eg SetLocalForward). So OnValidate is only necessary if you're tweaking the values by hand in the editor.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/ParabolicTeleportPointer.cs (hunk 5 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/ParabolicTeleportPointer.cs]: Teleportation and tooltips work - if you know of other components using parabolas let me know and I'll test them as well.
